torchdata/dataloader2/graph/utils.py

```diff
@@ -30,21 +30,31 @@ def helper(g) -> None:  # pyre-ignore
     return dps
 
 
-def list_dps(graph: DataPipeGraph, exclude_dps: Optional[Union[DataPipe, List[DataPipe]]] = None) -> List[DataPipe]:
+def list_dps(
+    graph: DataPipeGraph, exclude_dps: Optional[Union[DataPipe, List[DataPipe]]] = None, exclusive=True
+) -> List[DataPipe]:
     r"""
     Given the graph of DataPipe generated by ``traverse_dps`` function, return a list
     of all DataPipe instances without duplication. If ``exclude_dps`` is provided,
-    the provided ``DataPipes`` and their predecessors will be ignored.
+    the provided ``DataPipes`` and their predecessors will be ignored. If ``exclusive``
+    is set to True (default), all of the predecessors will be excluded even though
+    they are prior to any non-excluded ``DataPipe`` on different branches in the grpah.
     """
     dps: List[DataPipe] = []
     cache: Set[int] = set()
 
     if exclude_dps is not None:
         if isinstance(exclude_dps, (IterDataPipe, MapDataPipe)):
-            cache.add(id(exclude_dps))
-        else:
-            for dp in exclude_dps:  # type: ignore[union-attr]
-                cache.add(id(dp))
+            exclude_dps = [
+                exclude_dps,
+            ]
+        for exclude_dp in exclude_dps:  # type: ignore[union-attr]
+            assert isinstance(exclude_dp, (IterDataPipe, MapDataPipe))
+            if exclusive:
+                for dp in list_dps(traverse_dps(exclude_dp)):  # type: ignore[arg-type]
+                    cache.add(id(dp))
+            else:
+                cache.add(id(exclude_dp))
```

Comment:
Before this PR, `cache` would only store the id of `exclude_dps`. However, there are scenarios like:
```mermaid
graph TD;
DP1-->DP2;
DP2-->DP3;
DP1-->DP3;
DP3-->output;
```
If we exclude `DP2`, `DP1` will still be included in the result as `[DP1, DP3]`. After this PR:
- `exclusive=True`: Return `[DP3]`
- `exclusive=False`: Return `[DP1, DP3]` (previous behavior)

BTW, for the argument name, appreciate any recommendation based on the behavior.

Reply:
Just change the behavior to `exclusive=True` without adding the arg and update the docstring/example.